tests/integration/test_charm.py

```diff
@@ -68,39 +69,54 @@ async def test_relations(self, ops_test: OpsTest):
             timeout=600,
         )
 
-    # uncomment when juju 2.9.43 is released
-    # @pytest.mark.abort_on_fail
-    # @pytest.mark.timeout(1200)
-    # async def test_upgrade(self, ops_test: OpsTest):
-    #     await ops_test.model.remove_application(APP_NAME, block_until_done=True)
-    #
-    #     await ops_test.model.deploy(
-    #         APP_NAME, channel="ckf-1.7/stable", trust=True, config=OIDC_CONFIG
-    #     )
-    #     await ops_test.model.add_relation(f"{ISTIO_PILOT}:ingress", f"{APP_NAME}:ingress")
-    #     await ops_test.model.add_relation(
-    #         f"{ISTIO_PILOT}:ingress-auth", f"{APP_NAME}:ingress-auth"
-    #     )
-    #     await ops_test.model.add_relation(f"{APP_NAME}:oidc-client", f"{DEX_AUTH}:oidc-client")
-    #     await ops_test.model.applications[APP_NAME].set_config({"public-url": PUBLIC_URL})
-    #     await ops_test.model.wait_for_idle(
-    #         [APP_NAME, ISTIO_PILOT, DEX_AUTH],
-    #         status="active",
-    #         raise_on_blocked=False,
-    #         raise_on_error=True,
-    #         timeout=600,
-    #     )
-    #
-    #     cmd = (
-    #         f"juju refresh {APP_NAME} "
-    #         f'--path "{pytest.charm_under_test}" --resource oci-image="{image_path}"'
-    #     )
-    #     await ops_test.run(*shlex.split(cmd))
-    #
-    #     await ops_test.model.wait_for_idle(
-    #         [APP_NAME, ISTIO_PILOT, DEX_AUTH],
-    #         status="active",
-    #         raise_on_blocked=True,
-    #         raise_on_error=True,
-    #         timeout=600,
-    #     )
+    @pytest.mark.abort_on_fail
+    @pytest.mark.timeout(1200)
+    async def test_upgrade(self, ops_test: OpsTest):
+        """Test that charm can be upgraded from podspec to sidecar.
+
+        For this test we use 1.7/stable channel as the source for podspec charm.
+
+        Note: juju has a bug due to which you have to first scale podspec charm to 0,
+        then refresh, then scale up newly deployed app.
+        See https://github.com/juju/juju/pull/15701 for more info.
+        """
+        await ops_test.model.remove_application(APP_NAME, block_until_done=True)
```

Comment:
no, since we want to test refresh from stable to newly built charm

and newly built charm is already present, so, we remove it, install stable, refresh to edge

it could be avoided by chanding the order of tests. But if refresh fails. then we run our tests on stable charm, and get false positive

Reply:
If we're expecting the remove to fail, perhaps we should add a distinct test for that just before the upgrade. Since the Gatekeeper charm does not come with special `_on_remove` functionality, it could be as simple as:

```python
@pytest.mark.abort_on_fail
async def test_remove_application(self, ops_test: OpsTest):
    """Test that the application can be removed successfully."""
    await ops_test.model.remove_application(APP_NAME, block_until_done=True)
```

But the rationale is still the same, we should abstract setup/configuration logic from the tests themselves and strive for having tests that check specific things (e.g. either app removal **or** upgrade).

Reply:
that makes sense, let me fire a commit